File: test/methylome_test.cpp
/* MIT License
 *
 * Copyright (c) 2024 Andrew D Smith
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <methylome.hpp>

#include <cpg_index.hpp>
#include <methylome_data.hpp>

#include <gtest/gtest.h>

#include <cerrno>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <random>
#include <string>

using namespace xfrase;  // NOLINT

[[nodiscard]] auto
generate_unique_dir_name() -> std::string {
  // Generate a random string based on current time and random numbers
  auto now = std::chrono::system_clock::now().time_since_epoch().count();
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_int_distribution<> dis(1000, 9999);

  return "/tmp/test_dir_" + std::to_string(now) + "_" +
         std::to_string(dis(gen));
}

[[nodiscard]] auto
change_permissions_to_no_write(const std::filesystem::path &dir)
  -> std::error_code {
  std::error_code ec;
  // clang-format off
  namespace fs = std::filesystem;
  fs::permissions(dir, (fs::perms::owner_read |
                        fs::perms::owner_exec |
                        fs::perms::group_read |
                        fs::perms::group_exec |
                        fs::perms::others_read),
                  fs::perm_options::replace, ec);
  // clang-format on
  return ec;
}

[[nodiscard]] auto
write_should_fail(const std::filesystem::path &dir,
                  std::error_code &ec) -> bool {
  bool write_failed = false;
  std::filesystem::path file = dir / "test_file.txt";
  std::ofstream test_file(file);
  if (!test_file.is_open()) {
    ec = std::make_error_code(std::errc(errno));
    write_failed = true;
  }
  if (!write_failed) {
    test_file << "Test content";
    if (test_file.fail()) {
      ec = std::make_error_code(std::errc(errno));
      write_failed = true;
    }
    std::error_code local_ec;
    const auto file_exists = std::filesystem::exists(file, local_ec);
    if (file_exists)
      std::filesystem::remove(file, local_ec);
  }
  return write_failed;
}

TEST(methylome_test, invalid_accession) {
  auto res = methylome::is_valid_name("invalid.accession");
  EXPECT_FALSE(res);
  res = methylome::is_valid_name("invalid/accession");
  EXPECT_FALSE(res);
}

TEST(methylome_test, valid_accessions) {
  auto res = methylome::is_valid_name("eFlareon_brain");
  EXPECT_TRUE(res);
  res = methylome::is_valid_name("SRX012345");
  EXPECT_TRUE(res);
}

TEST(methylome_test, valid_read) {
  static constexpr auto methylome_directory = "data/lutions/methylomes";
  static constexpr auto methylome_name = "eJolteon_ear";
  std::error_code ec;
  [[maybe_unused]] const auto meth =
    methylome::read(methylome_directory, methylome_name, ec);
  EXPECT_FALSE(ec);
}

TEST(methylome_test, invalid_read) {
  static constexpr auto methylome_directory = "data/lutions/methylomes";
  static constexpr auto methylome_name = "eVolteon_ear";
  std::error_code ec;
  [[maybe_unused]] const auto meth =
    methylome::read(methylome_directory, methylome_name, ec);
  EXPECT_EQ(ec, std::errc::no_such_file_or_directory);
}

TEST(methylome_test, is_consistent) {
  static constexpr auto methylome_directory = "data/lutions/methylomes";
  static constexpr auto methylome_name = "eFlareon_brain";
  std::error_code ec;
  const auto meth = methylome::read(methylome_directory, methylome_name, ec);
  EXPECT_FALSE(ec);
  EXPECT_TRUE(meth.is_consistent());
}

TEST(methylome_test, valid_write) {
  static constexpr auto methylome_directory = "data/lutions/methylomes";
  static constexpr auto output_directory = "data/lutions";
  static constexpr auto methylome_name = "eFlareon_brain";
  std::error_code ec;
  const auto meth = methylome::read(methylome_directory, methylome_name, ec);
  EXPECT_FALSE(ec);

  ec = meth.write(output_directory, methylome_name);
  EXPECT_FALSE(ec);

  const auto meta_filename =
    methylome_metadata::compose_filename(output_directory, methylome_name);
  const auto meta_file_exists = std::filesystem::exists(meta_filename, ec);
  EXPECT_FALSE(ec);
  EXPECT_TRUE(meta_file_exists);
  if (meta_file_exists) {
    std::filesystem::remove(meta_filename, ec);
    EXPECT_FALSE(ec);
  }

  const auto data_filename =
    methylome_data::compose_filename(output_directory, methylome_name);
  const auto data_file_exists = std::filesystem::exists(data_filename, ec);
  EXPECT_FALSE(ec);
  EXPECT_TRUE(data_file_exists);
  if (data_file_exists) {
    std::filesystem::remove(data_filename, ec);
    EXPECT_FALSE(ec);
  }
}

TEST(methylome_test, invalid_write) {
  static constexpr auto methylome_directory = "data/lutions/methylomes";
  static constexpr auto methylome_name = "eFlareon_brain";

  std::error_code ec;
  const auto output_directory = generate_unique_dir_name();
  std::filesystem::create_directory(output_directory, ec);
  if (ec)
    return;

  ec = change_permissions_to_no_write(output_directory);
  if (ec)
    return;

  const auto proceed_this_test = write_should_fail(output_directory, ec);
  if (ec || !proceed_this_test)
    return;

  const auto meth = methylome::read(methylome_directory, methylome_name, ec);
  EXPECT_FALSE(ec);

  ec = meth.write(output_directory, methylome_name);
  EXPECT_EQ(ec, std::errc::permission_denied);

  const auto meta_filename =
    methylome_metadata::compose_filename(output_directory, methylome_name);
  const auto meta_file_exists = std::filesystem::exists(meta_filename, ec);
  EXPECT_FALSE(ec);

  EXPECT_FALSE(meta_file_exists);
  if (meta_file_exists) {
    std::filesystem::remove(meta_filename, ec);
    EXPECT_FALSE(ec);
  }

  const auto data_filename =
    methylome_data::compose_filename(output_directory, methylome_name);
  const auto data_file_exists = std::filesystem::exists(data_filename, ec);
  EXPECT_FALSE(ec);
  EXPECT_FALSE(data_file_exists);
  if (data_file_exists) {
    std::filesystem::remove(data_filename, ec);
    EXPECT_FALSE(ec);
  }

  std::filesystem::remove_all(output_directory, ec);
  EXPECT_FALSE(ec);
}

TEST(methylome_test, init_metadata) {
  static constexpr auto methdir = "data/lutions/methylomes";
  static constexpr auto indexdir = "data/lutions/indexes";
  static constexpr auto methylome_name = "eVaporeon_tail";
  static constexpr auto genome_name = "eVaporeon";

  std::error_code ec;
  const auto index = cpg_index::read(indexdir, genome_name, ec);
  EXPECT_FALSE(ec);

  const auto meta = methylome_metadata::read(methdir, methylome_name, ec);
  EXPECT_FALSE(ec);

  const auto data = methylome_data::read(methdir, methylome_name, meta, ec);
  EXPECT_FALSE(ec);

  methylome meth{data, meta};

  ec = meth.init_metadata(index);
  EXPECT_FALSE(ec);
  EXPECT_TRUE(meth.is_consistent());
}

TEST(methylome_test, update_metadata) {
  static constexpr auto methdir = "data/lutions/methylomes";
  static constexpr auto methylome_name = "eVaporeon_tail";

  std::error_code ec;
  const auto meta = methylome_metadata::read(methdir, methylome_name, ec);
  EXPECT_FALSE(ec);

  const auto data = methylome_data::read(methdir, methylome_name, meta, ec);
  EXPECT_FALSE(ec);

  methylome meth{data, meta};

  ec = meth.update_metadata();
  EXPECT_FALSE(ec);
  EXPECT_TRUE(meth.is_consistent());
}

File: python/test/methylome_test.py
# MIT License
#
# Copyright (c) 2024 Andrew D Smith
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


import pytest
import shutil
import tempfile
import os

from transferase import GenomeIndex
from transferase import Methylome
from transferase import GenomicInterval
from transferase import MLevels
from transferase import MLevelsCovered


def get_valid_test_genome_index(genome_name):
    """
    Fixture to load a valid GenomeIndex object for testing
    """
    genome_index_dirname = "data/lutions/indexes"
    genome_index = GenomeIndex.read(genome_index_dirname, genome_name)
    return genome_index


def get_valid_test_methylome(genome_name, tissue_name):
    """
    Fixture to load a valid Methylome object for testing
    """
    methylome_directory = "data/lutions/methylomes"
    methylome_name = f"{genome_name}_{tissue_name}"
    meth = Methylome.read(methylome_directory, methylome_name)
    return meth


def get_valid_test_genomic_intervals(genome_name, tissue_name):
    """
    Fixture to load a valid list of GenomicInterval objects for testing
    """
    intervals_directory = "data/lutions/raw"
    intervals_filename = os.path.join(
        intervals_directory,
        f"{genome_name}_{tissue_name}_hmr.bed",
    )
    genome_index = get_valid_test_genome_index(genome_name)
    intervals = GenomicInterval.read(genome_index, intervals_filename)
    return intervals


def get_valid_test_query(genome_name, tissue_name):
    """
    Fixture to make a valid query for given GenomeIndex and a
    corresponding list of GenomicInterval objects
    """
    intervals = get_valid_test_genomic_intervals(genome_name, tissue_name)
    genome_index = get_valid_test_genome_index(genome_name)
    query = genome_index.make_query(intervals)
    assert query is not None, f"Failed to make query from fixture"
    return query


def create_temp_directory():
    """Create a unique temporary directory in /tmp"""
    temp_dir = tempfile.mkdtemp(dir="/tmp", prefix="test_")
    return temp_dir


def test_methylome_init():
    """Test the default constructor"""
    obj = Methylome()
    assert obj is not None


def test_methylome_read():
    """Test the read static method"""
    directory = "data/lutions/methylomes"
    methylome_name = "eFlareon_brain"
    meth = Methylome.read(directory, methylome_name)
    assert meth is not None


def test_methylome_is_consistent():
    """Test the is_consistent method (empty Methylome)"""
    meth = Methylome()
    assert isinstance(meth.is_consistent(), bool)


def test_methylome_is_consistent():
    """Test the is_consistent method (non-empty Methylome)"""
    directory = "data/lutions/methylomes"
    methylome_name = "eFlareon_brain"
    meth = Methylome.read(directory, methylome_name)
    assert meth.is_consistent()


def test_methylome_is_consistent_with_other():
    """Test the is_consistent method (with another methylome object)"""
    obj1 = Methylome()
    obj2 = Methylome()
    assert isinstance(obj1.is_consistent(obj2), bool)


def test_methylome_write():
    """Test the write method"""
    genome_name = "eVaporeon"
    tissue_name = "brain"
    methylome_name = f"{genome_name}_{tissue_name}"
    meth = get_valid_test_methylome(genome_name, tissue_name)
    output_directory = create_temp_directory()
    meth.write(output_directory, methylome_name)
    if os.path.isdir(output_directory):
        shutil.rmtree(output_directory)


def test_methylome_init_metadata_inconsistent():
    """
    Test init_metadata method when Methylome and GenomeIndex are not
    consistent
    """
    genome_name1 = "eVaporeon"
    genome_name2 = "eJolteon"
    tissue_name = "ear"
    index = get_valid_test_genome_index(genome_name1)
    meth = get_valid_test_methylome(genome_name2, tissue_name)
    with pytest.raises(RuntimeError, match="invalid methylome data"):
        meth.init_metadata(index)


def test_methylome_init_metadata_consistent():
    """
    Test init_metadata method when Methylome and GenomeIndex are
    consistent
    """
    genome_name = "eFlareon"
    tissue_name = "tail"
    index = get_valid_test_genome_index(genome_name)
    meth = get_valid_test_methylome(genome_name, tissue_name)
    try:
        meth.init_metadata(index)
    except RuntimeError as run_err:
        pytest.fail(f"Unexpected exception raised: {run_err}")


def test_methylome_update_metadata_empty():
    """Test update_metadata method for an empty Methylome"""
    meth = Methylome()
    try:
        meth.update_metadata()
    except RuntimeError as run_err:
        pytest.fail(f"Unexpected exception raised: {run_err}")


def test_methylome_update_metadata():
    """Test update_metadata method"""
    genome_name = "eVaporeon"
    tissue_name = "brain"
    meth = get_valid_test_methylome("eVaporeon", "brain")
    try:
        meth.update_metadata()
    except RuntimeError as run_err:
        pytest.fail(f"Unexpected exception raised: {run_err}")


def test_methylome_add_empty():
    """Test add method for two empty methylomes"""
    meth1 = get_valid_test_methylome("eJolteon", "brain")
    meth2 = get_valid_test_methylome("eJolteon", "ear")
    meth1.add(meth2)
    assert meth1 is not None


def test_methylome_repr():
    """Test __repr__ method"""
    meth = Methylome()
    repr_result = repr(meth)
    assert isinstance(repr_result, str)
    assert "version" in repr_result


def test_methylome_get_levels_with_query_container():
    """Test get_levels method (with query_container argument)"""
    meth = get_valid_test_methylome("eFlareon", "brain")
    query = get_valid_test_query("eFlareon", "brain")
    levels = meth.get_levels(query)
    assert isinstance(levels, MLevels)


def test_methylome_get_levels_covered_with_query_container():
    """Test get_levels_covered method (with query_container argument)"""
    meth = get_valid_test_methylome("eJolteon", "ear")
    query = get_valid_test_query("eJolteon", "ear")
    levels = meth.get_levels_covered(query)
    assert isinstance(levels, MLevelsCovered)


def test_methylome_get_levels_with_bin_size_and_genome_index():
    """Test get_levels method (with bin_size and genome_index argument)"""
    genome_name = "eJolteon"
    tissue_name = "ear"
    meth = get_valid_test_methylome(genome_name, tissue_name)
    genome_index = get_valid_test_genome_index(genome_name)
    bin_size = 100
    levels = meth.get_levels(bin_size, genome_index)
    assert isinstance(levels, MLevels)


def test_methylome_get_levels_covered_with_bin_size_and_genome_index():
    """Test get_levels_covered method (with bin_size and genome_index argument)"""
    genome_name = "eJolteon"
    tissue_name = "ear"
    meth = get_valid_test_methylome(genome_name, tissue_name)
    genome_index = get_valid_test_genome_index(genome_name)
    bin_size = 100
    levels = meth.get_levels_covered(bin_size, genome_index)
    assert isinstance(levels, MLevelsCovered)


def test_methylome_global_levels():
    """Test global_levels method"""
    meth = Methylome()
    result = meth.global_levels()
    assert isinstance(result, tuple), "result should be a tuple"
    assert len(result) == 2, "result should have 2 elements"
    assert result[0] == 0


def test_methylome_global_levels_covered():
    """Test global_levels_covered method"""
    meth = Methylome()
    result = meth.global_levels_covered()
    assert isinstance(result, tuple), "result should be a tuple"
    assert len(result) == 3, "result should have 3 elements"
    assert result[2] == 0
